dipy/denoise/enhancement_kernel.pyx

```diff
@@ -0,0 +1,420 @@
+                   double [:] r, double [:] v) nogil:
+        """ Evaluate the kernel at position x relative to
+        position y, with orientation r relative to orientation v.
+        """
+        cdef:
+            double [:] a
+            double [:,:] transm
+            double [:] arg1
+            double [:] arg2p
+            double [:] arg2
+            double [:] c
+            double kernelval
+
+        with gil:
+
+            a = np.subtract(x, y)
+            transm = np.transpose(R(euler_angles(v)))
+            arg1 = np.dot(transm, a)
+            arg2p = np.dot(transm, r)
+
+        arg2 = euler_angles(arg2p)
+
+        c = self.coordinate_map(arg1[0], arg1[1], arg1[2],
+                                arg2[0], arg2[1])
+        kernelval = self.kernel(c)
+
+        return kernelval
+
+    @cython.wraparound(False)
+    @cython.boundscheck(False)
+    @cython.nonecheck(False)
+    @cython.cdivision(True)
+    cdef double [:] coordinate_map(self, double x, double y,
+                                   double z, double beta,
+                                   double gamma) nogil:
+        """ Compute a coordinate map for the kernel
+
+        Parameters
+        ----------
+        x : double
+            X position
+        y : double
+            Y position
+        z : double
+            Z position
+        beta : double
+            First Euler angle
+        gamma : double
+            Second Euler angle
+
+        Returns
+        -------
+        c : array of double
+            array of coordinates for kernel
+        """
+
+        cdef:
+            double [:] c
+            double q
+            double cg
+            double cotq2
+
+        with gil:
+
+            c = np.zeros(6)
+
+        if beta == 0:
+            c[0] = x
+            c[1] = y
+            c[2] = z
+            c[3] = c[4] = c[5] = 0
+
+        else:
+            q = fabs(beta)
+            cg = cos(gamma)
+            sg = sin(gamma)
+            cotq2 = 1.0 / tan(q/2)
+
+            c[0] = -0.5*z*beta*cg + \
+                    x*(1 - (beta*beta*cg*cg * (1 - 0.5*q*cotq2)) / (q*q)) - \
+                    (y*beta*beta*cg*sg * (1 - 0.5*q*cotq2)) / (q*q)
+            c[1] = -0.5*z*beta*sg - \
+                    (x*beta*beta*cg*sg * (1 - 0.5*q*cotq2)) / (q*q) + \
+                    y * (1 - (beta*beta*sg*sg * (1 - 0.5*q*cotq2)) / (q*q))
+            c[2] = 0.5*x*beta*cg + 0.5*y*beta*sg + \
+                   z * (1 + ((1 - 0.5*q*cotq2) * (-beta*beta*cg*cg - \
+                        beta*beta*sg*sg)) / (q*q))
+            c[3] = beta * (-sg)
+            c[4] = beta * cg
+            c[5] = 0
+
+        with gil:
+
+            return np.array(c)
+
+    @cython.wraparound(False)
+    @cython.boundscheck(False)
+    @cython.nonecheck(False)
+    @cython.cdivision(True)
+    cdef double kernel(self, double [:] c) nogil:
+        """ Evaluate the kernel based on the coordinate map.
+        """
+        return 1 / (8*sqrt(2)) * \
```

Comment:
Would it be too much to break this equation up into sub-parts, to increase it's readability? It's pretty hard to see the math as it is right now.